File: crates/embroiderly-pattern/src/publish.rs
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "borsh", derive(borsh::BorshSerialize, borsh::BorshDeserialize))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct PublishSettings {
  pub pdf: PdfExportOptions,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "borsh", derive(borsh::BorshSerialize, borsh::BorshDeserialize))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct PdfExportOptions {
  /// Maximum size of a frame in stitches (width, height).
  pub frame_size: (u16, u16),
  /// Number of overlapping rows/columns from adjacent frames to include.
  pub preserved_overlap: u16,
  /// Whether to show grid line numbers in the exported PDF.
  pub show_grid_line_numbers: bool,
  /// Whether to show centering marks in the exported PDF.
  pub show_centering_marks: bool,
}

impl Default for PdfExportOptions {
  fn default() -> Self {
    Self {
      frame_size: (30, 40),
      preserved_overlap: 3,
      show_grid_line_numbers: true,
      show_centering_marks: true,
    }
  }
}

File: app/src/lib/pattern/publish.ts
import { b } from "@zorsh/zorsh";

export class PdfExportOptions {
  frameSize: [number, number];
  preservedOverlap: number;
  showGridLineNumbers: boolean;
  showCenteringMarks: boolean;

  constructor(data?: Partial<b.infer<typeof PdfExportOptions.schema>>) {
    this.frameSize = data?.frameSize ?? [30, 40];
    this.preservedOverlap = data?.preservedOverlap ?? 3;
    this.showGridLineNumbers = data?.showGridLineNumbers ?? true;
    this.showCenteringMarks = data?.showCenteringMarks ?? true;
  }

  static readonly schema = b.struct({
    frameSize: b.tuple(b.u16(), b.u16()),
    preservedOverlap: b.u16(),
    showGridLineNumbers: b.bool(),
    showCenteringMarks: b.bool(),
  });

  static deserialize(data: Uint8Array) {
    return new PdfExportOptions(PdfExportOptions.schema.deserialize(data));
  }

  static serialize(data: PdfExportOptions) {
    return PdfExportOptions.schema.serialize(data);
  }
}

export class PublishSettings {
  pdf: PdfExportOptions;

  constructor(data?: Partial<b.infer<typeof PublishSettings.schema>>) {
    this.pdf = new PdfExportOptions(data?.pdf);
  }

  static readonly schema = b.struct({
    pdf: PdfExportOptions.schema,
  });
}
